submissions/factorio_draftsman/main.md

```diff
@@ -0,0 +1,211 @@
+
+```python
+# Create a N x 5 grid of connected constant combinators, with every item and their stack size
+
+from draftsman.blueprintable import Blueprint
+from draftsman.constants import Direction
+from draftsman.data import items
+from draftsman.entity import ConstantCombinator
+
+COMBINATOR_HEIGHT = 5
+
+def main():
+    blueprint = Blueprint()
+
+    signal_count = 0
+    signal_index = 0
+    combinators_added = 0
+    x = 0
+    y = 0
+    combinator = ConstantCombinator(direction=Direction.SOUTH)
+
+    # Iterate over every item in order:
+    for item in items.raw:
+        # Ignore hidden items/entities
+        if "flags" in items.raw[item]:
+            if "hidden" in items.raw[item]["flags"]:
+                continue
+        # Keep track of how many signals we've gone through
+        signal_count += 1
+        # Write the stack size signal
+        stack_size = items.raw[item]["stack_size"]
+        combinator.set_signal(signal_index, item, stack_size)
+        signal_index += 1
+        # Once we exceed the number of signals a combinator can hold, place it and reset
+        if signal_index == combinator.item_slot_count:
+            # Add the combinator to the blueprint
+            combinator.id = "{}_{}".format(x, y)
+            blueprint.entities.append(combinator)
+            # Reset the combinator
+            combinators_added += 1
+            y = combinators_added % COMBINATOR_HEIGHT
+            x = int(combinators_added / COMBINATOR_HEIGHT)
+            combinator.set_signals(None)  # Clear signals
+            combinator.tile_position = (x, y)
+            signal_index = 0
+
+    # Add the last combinator if partially full
+    combinator.id = "{}_{}".format(x, y)
+    blueprint.entities.append(combinator)
+
+    # Add connections to each neighbour
+    for cx in range(x):
+        for cy in range(COMBINATOR_HEIGHT):
+            here = "{}_{}".format(cx, cy)
+            right = "{}_{}".format(cx + 1, cy)
+            below = "{}_{}".format(cx, cy + 1)
+            try:
+                blueprint.add_circuit_connection("red", here, right)
+            except KeyError:
+                pass
+            try:
+                blueprint.add_circuit_connection("red", here, below)
+            except KeyError:
+                pass
+
+    print("Number of items added:", signal_count)
```

Comment:
Maybe rename `signal_count` to `item_count`, or print `len(items.raw)` here so you can remove the variable altogether?

Reply:
The `item_count` vs. `signal_count` was something I debated over changing; basically it depends on what you're looking at. If you're looking at it from the perspective of the items, then `item_count` makes more sense, but if you consider the fact that all of these eventually get resolved to constant combinator signals, `signal_count` makes more sense. 

The "hidden" check also complicates this somewhat. I'm a bit fuzzy on it at the moment, but I seem to remember it being possible that items could be hidden from the user, or belonged to entities that were hidden from the user; these items are useless to a script like this because the end user is only concerned with constructible entities. Thus, it makes sense to iterate over all the items, ignore the ones that are hidden, and then add everything else as signals to the constant combinators. This way of thinking also supports the use of `signal_count` over `item_count`, because the total number of signals in the resulting blueprint is not necessarily the total number of items in `items.raw`, which `item_count` could be conflated with.

In order to truly answer the question, I need to research if my assumption that items can be hidden is true, or if that was an assumption I made early on that I forgot to rigorously check; this was one of the first functioning scripts I wrote, so it's likely to have some holdovers from previous iterations. Let me get back to you on this.

Reply:
Alright, it is necessary. Using the following code snippet on a vanilla install of Draftsman:
```python
for item_name in items.raw:
    item = items.raw[item_name]
    if "flags" in item and "hidden" in item["flags"]:
        print(item_name)
```
We get the following list:
```
loader
fast-loader
express-loader
rocket-part
coin
tank-machine-gun
vehicle-machine-gun
tank-flamethrower
artillery-wagon-cannon
spidertron-rocket-launcher-1
spidertron-rocket-launcher-2
spidertron-rocket-launcher-3
spidertron-rocket-launcher-4
tank-cannon
player-port
item-unknown
electric-energy-interface
linked-chest
heat-interface
linked-belt
infinity-chest
infinity-pipe
simple-entity-with-force
simple-entity-with-owner
burner-generator
```
A lot of these are just special entities that only exist in the editor, like infinity chests and loaders. Some are special components of larger entities, like the spidertron rocket launchers, and coin is a regular item that just isn't used. Thus, we do have to check each item to see if it's "valid" before proceeding.

I propose the following changes to make this more clear:
* Rename `signal_count` to `signals_added`, in the same vein as `combinators_added`
* Move `signals_added += 1` to after we set the signal to make it a little more procedural
* Change the final `print("Number of items added:", signals_added)` to `print("Total number of item signals added:", signals_added)` to make it more clear